add rubocop and drop support for Rubies < 2.0

hedgesky opened the pull request:
This PR consists of two commits. First one introduces following changes:

- Adds `rubocop` as a dev dependency (use conservative version locking as official manual [advises](http://rubocop.readthedocs.io/en/latest/installation/);
- Adds `.rubocop.yml` config;
- Fixes some style issues;
- Changes default `rake` task;
- Updates `travis.yml` to run `rake` instead of `rspec`

When first commit failed Travis checks I realized that fresh Rubocop requires only Ruby >= 2.0. Keeping in mind that [there is a plan to drop support for Rubies < 2.1](https://github.com/collectiveidea/interactor/projects/1) in Interactor 4.0, I've decided to update this PR accordingly.

So the second commit does following: 
- removes Rubies 1.9.3, 2.0 from Travis config;
- sets Rubocop target Ruby version to 2.1;
- add `required_ruby_version = ">= 2.1"` in gemspec.

laserlemon commented on spec/integration_spec.rb:
```diff
@@ -25,7 +25,8 @@ def build_organizer(options = {}, &block)
   let(:organizer) {
-    build_organizer(organize: [organizer2, interactor3, organizer4, interactor5]) do
+    interactors = [organizer2, interactor3, organizer4, interactor5]
+    build_organizer(organize: interactors) do
```
Are this and the similar changes below just a line length concern?

hedgesky replied:
Yep. Both ways have their benefits and downsides. Introducing intermediate variable seems a bit clumsy, but it keeps lines short. Short lines are easier to perceive and they are two-panes-friendly:
![2017-03-22_15-36-32](https://cloud.githubusercontent.com/assets/3628124/24198083/5e966b4c-0f15-11e7-908b-d9ac070a9354.png)

laserlemon replied:
Yeah, I'm all for the 80 character limit but I typically relax that rule a bit for specs. But maybe I shouldn't!

laserlemon commented on spec/integration_spec.rb:
```diff
@@ -345,12 +347,10 @@ def rollback
           raise "foo"
-          context.steps << :around_before
-          interactor.call
-          context.steps << :around_after
```
Although it may seem obvious to Rubocop, there's some value in the assertion that these lines are _not_ called, even after a `raise`. This applies to several of the changes below as well.

laserlemon replied:
I'm open to discussion on this point!

hedgesky replied:
Yeah, I see your point. I thought about smth like this:

```ruby
def unexpected_error!
  raise "foo"
end
```

It may be more intention-revealing (the "unexpected" nature), but plain `raise` is far simpler and understable. What do you think?

laserlemon replied:
I like your use of `unexpected_error!` although we might have to define that method within the interactor's definition in the spec rather than as a helper method in the spec, which muddies the waters a little bit.

laserlemon commented on spec/support/lint.rb:
```diff
@@ -43,7 +43,9 @@
-      expect(Interactor::Context).to receive(:build).once.with(foo: "bar") { context }
+      expect(
+        Interactor::Context
+      ).to receive(:build).once.with(foo: "bar") { context }
```
If this is a line length concern, would Rubocop accept this style:

```ruby
expect(Interactor::Context).to receive(:build)
  .once.with(foo: "bar") { context }
```

hedgesky replied:
Sure!

laserlemon commented on interactor.gemspec:
```diff
@@ -1,19 +1,24 @@
 # encoding: utf-8
+require "English"
```
Thank you, I didn't know about this!

laserlemon commented on interactor.gemspec:
```diff
@@ -1,19 +1,24 @@
+  spec.files      = `git ls-files`.split($INPUT_RECORD_SEPARATOR)
   spec.test_files = spec.files.grep(/^spec/)
 
+  spec.required_ruby_version = ">= 2.1"
```
👏 🏆 💐 🥇

laserlemon commented on interactor.gemspec:
```diff
@@ -1,19 +1,24 @@
+
   spec.add_development_dependency "bundler", "~> 1.7"
   spec.add_development_dependency "rake", "~> 10.3"
+  spec.add_development_dependency "rubocop", "~> 0.47.1"
```
I'd rather put this dependency in the `Gemfile` and keep the gemspec dependencies to only `bundler` and `rake`.

laserlemon commented on lib/interactor.rb:
```diff
@@ -153,14 +153,12 @@ def run!
   # Returns nothing.
-  def call
-  end
+  def call; end
```
I'm not a big fan of this, but I suppose if the community has spoken! 😄

hedgesky replied:
Rubocop default settings are easily adjustable to project's needs. I'll remove this change.

laserlemon commented:
Actually, this may need to be reopened as a PR against the `v4` branch as well. 😕

hedgesky commented:
I've added replies to your comments; some of them require further discussion. I'll close this PR and reopen a new one on `v4` branch.

hedgesky commented:
Actually, it's possible to just change base branch via github UI. Closed this too rashly 😄

laserlemon commented:
I used the GitHub web UI to resolve conflicts with the `v4` branch. I may have inadvertently removed the Ruby version restriction from the gemspec, but on second thought, I'm not sure that we want it. While we're dropping support for Rubies < 2.1, it's possible that Interactor 4.0 _could_ work with Ruby 2.0, depending on what Ruby features we take advantage of. Dropping support for Ruby 2.0 is more of a statement that you can _try_ to use Interactor with Ruby 2.0 but do so at your own risk. I'm open to discussion on this point though. It's certainly an option to draw a harder line with our Ruby version restriction in the gemspec.

hedgesky commented:
Changes in last commit:

- I've decided to keep 80 character limit in `integration_spec`;
- Added `#unexpected_error!` in `integration_spec`;
- Changed `spec/support/lint.rb:46` following your request;
- Set `spec.required_ruby_version = ">= 2.0"`;
- Moved rubocop from gemspec to Gemfile;
- Revert change with one-lined empty `def`s in `lib/interactor`.

About Ruby versions: support for required keyword arguments appeared only in 2.1. I think we'll definitely need them if we're going to implement #123.

laserlemon commented:
@hedgesky Your changes look really good. You said in your comment that keyword arguments appeared in Ruby 2.0, but I'm seeing [otherwise](https://robots.thoughtbot.com/ruby-2-keyword-arguments). You're absolutely right though that those will most likely be required so that would be the right reason to scope the Ruby version in the gemspec. It seems like we'll need to scope it to `>= 2.0` and then only officially support non-EOL Rubies, which would be `>= 2.1` as of this writing. Does that sound right to you?

hedgesky commented:
Yeah, keyword arguments appeared in 2.0, but I told about required ones. They appeared in 2.1. Both your link and [official changelog](https://github.com/ruby/ruby/blob/v2_1_0/NEWS) on Ruby github mentions that.
Most likely we'll need both required and non-required arguments, so maybe 2.0 support doesn't make sense in our case.

laserlemon commented:
It could be that even though required keyword arguments would be a big win for Interactor 4, they might not be strictly necessary because the structure of the keyword arguments is up to the developer when they define the `call` instance method. Plus, I imagine defining keyword arguments at all will be completely optional. There may be some difference between Rubies 2.0 and 2.1 as far as how we determine the nature of what keyword arguments are being requested so we can extract those from the context and send them along.

hedgesky commented:
I suggest delaying the decision about 2.0/2.1 until we have more information about Interactor 4 implementation details. For now we can just stick with 2.0.
Is it OK with you?

laserlemon commented:
@hedgesky Yes, that sounds fine, but let's make one change then. If we'll target Ruby 2.0 for now, let's add 2.0 back to `.travis.yml` but allow it to fail, like we do for `ruby-head`, since it's EOL.

laserlemon commented:
@hedgesky Looks good. Is there anything else you want to add before I merge?

hedgesky commented:
Nope, that's all. Go right ahead 😄 
Thanks for the discussion, it was really productive! If there is anything else I can help with, feel free to reach out to me. I'll be glad to speed up Interactor 4 release ✈️

laserlemon commented:
@hedgesky Absolutely, I hope you don't feel like you're off the hook now! 😄 You've been a big help already and I know I'll need help to get 4.0 out the door. 🚀

hedgesky commented:
Looks like `Drop support for Rubies < 2.1` and `Add and configure Rubocop to ensure consistent code style` project cards can be moved to Done section now (really cool github feature, from my point of view).

I'm going to work on `raise -> throw` switch following #120. Would you mind if I reopen it? Probably it deserves  `v4.0.0` milestone, too.
